Add new systemd_info module

NomakCooper opened the pull request:
##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->
This PR adds a new module, **`systemd_info`**, to the **`community.general`** collection. 

I decided to develop the **`systemd_info`** module to give all Ansible users the ability to gather detailed information about systemd units ( **`service`**,**`target`**,**`socket`**,**`mount`**  ).

The module gathers detailed **`systemd`** unit ( **`service`**,**`target`**,**`socket`**,**`mount`**  ) info using the **`systemctl`** command. 

It supports both global and select modes, and includes the ability to collect additional properties as specified by the user.

### Key Features

- **Global and Select Modes:**
  - **Global mode:** Retrieves info for all systemd units using **`systemctl`**.
  - ```
      $ systemctl list-units --no-pager --type service,target,socket,mount --all --plain --no-legend
    ```
  - ```
      $ systemctl show 'UNIT_NAME' -p 'PROPERTIES_LIST'
    ```
  - **Select mode:** Processes only the units specified in the **`unitname`** parameter. The module first verifies unit existence and, if not found, module fail.
  
- **Customizable Properties:**
  - The `extra_properties` parameter allows users to request extra unit properties beyond the defaults.

- **Detailed Return Values:**
  - The user must use the **`register`** function to access the return values.

### Usage Examples

```yaml
# Gather info for all systemd services, targets, sockets and mount
- name: Gather all systemd unit info
  community.general.systemd_info:
  register: results

# Gather info for selected units with extra properties.
- name: Gather info for selected unit(s)
  community.general.systemd_info:
    unitname:
      - systemd-journald.service
      - systemd-journald.socket
      - sshd-keygen.target
      - -.mount
    extra_properties:
        - Description
    register: results
```
### Return Examples
```json
  "results.units": {
    "-.mount": {
        "activestate": "active",
        "description": "Root Mount",
        "loadstate": "loaded",
        "name": "-.mount",
        "options": "rw,relatime,seclabel,attr2,inode64,logbufs=8,logbsize=32k,noquota",
        "substate": "mounted",
        "type": "xfs",
        "what": "/dev/mapper/cs-root",
        "where": "/"
    },
    "sshd-keygen.target": {
        "activestate": "active",
        "description": "sshd-keygen.target",
        "fragmentpath": "/usr/lib/systemd/system/sshd-keygen.target",
        "loadstate": "loaded",
        "name": "sshd-keygen.target",
        "substate": "active",
        "unitfilepreset": "disabled",
        "unitfilestate": "static"
    },
    "systemd-journald.service": {
        "activestate": "active",
        "description": "Journal Service",
        "execmainpid": "613",
        "fragmentpath": "/usr/lib/systemd/system/systemd-journald.service",
        "loadstate": "loaded",
        "mainpid": "613",
        "name": "systemd-journald.service",
        "substate": "running",
        "unitfilepreset": "disabled",
        "unitfilestate": "static"
    },
    "systemd-journald.socket": {
        "activestate": "active",
        "description": "Journal Socket",
        "fragmentpath": "/usr/lib/systemd/system/systemd-journald.socket",
        "loadstate": "loaded",
        "name": "systemd-journald.socket",
        "substate": "running",
        "unitfilepreset": "disabled",
        "unitfilestate": "static"
    }
```
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->

<!--- Please do not forget to include a changelog fragment:
      https://docs.ansible.com/ansible/devel/community/collection_development_process.html#creating-changelog-fragments
      No need to include one for docs-only or test-only PR, and for new plugin/module PRs.
      Read about more details in CONTRIBUTING.md.
      -->

##### ISSUE TYPE
<!--- Pick one or more below and delete the rest.
      'Test Pull Request' is for PRs that add/extend tests without code changes. -->
- New Module/Plugin Pull Request

##### COMPONENT NAME
<!--- Write the SHORT NAME of the module, plugin, task or feature below. -->
systemd_info
##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here -->
### Base Properties table

| **Property**     | **.service**              | **.target**               | **.socket**               | **.mount**               |
|------------------|---------------------------|---------------------------|---------------------------|--------------------------|
| **name**         | Yes                       | Yes                       | Yes                       | Yes                      |
| **loadstate**    | Yes                       | Yes                       | Yes                       | Yes                      |
| **activestate**  | Yes                       | Yes                       | Yes                       | Yes                      |
| **substate**     | Yes                       | Yes                       | Yes                       | Yes                      |
| **fragmentpath** | Yes (if available)        | Yes (if available)        | Yes (if available)        | No                      |
| **unitfilestate**| Yes (if available)        | Yes (if available)        | Yes (if available)        | No                      |
| **unitfilepreset**| Yes (if available)       | Yes (if available)        | Yes (if available)        | No                      |
| **mainpid**      | Yes         | No                       | No                       | No                      |
| **execmainpid**  | Yes        | No                       | No                       | No                      |
| **Where**        | No                       | No                       | No                       | Yes                      |
| **What**         | No                       | No                       | No                       | Yes                      |
| **Options**      | No                       | No                       | No                       | Yes                      |
| **Type**         | No                       | No                       | No                       | Yes                      |

ALL possibile unit properties can be collected by user using **`unitname`**/**`extra_properties`** options

In this PR:

- Added the **`systemd_info`** module in **`plugins/modules/systemd_info.py`**.
- Added an integration test in **`tests/integration/targets/systemd_info/`**.
- Added an entry in **`.github/BOTMETA.yml`**.

Ansible sanity and integration tests were successfully run locally using Docker.
<!--- A step-by-step reproduction of the problem is helpful if there is no related issue -->
<!--- Paste verbatim command output below, e.g. before and after your change -->

NomakCooper commented:
I have corrected the result object because previously it was added directly to the **`register`**. 
This merged the **`changed`** and **`failed`** fields with the result, potentially causing misunderstandings or incorrect handling. Now, the module returns the result object within a **`units`** object inside the **`register`**.

like:
```yaml
results.units['sshd-keygen.target']
```

russoz reviewed (commented):
hi @NomakCooper 
Thanks for your PR! I've made an initial review.

NomakCooper commented:
@russoz 

Thanks for your review comments.

I've made some changes based on your suggestions:

- I improved the syntax of the condition in the integration test as you recommended.

    ```yaml
      when: >
        (ansible_distribution in ['RedHat', 'CentOS', 'ScientificLinux'] and ansible_distribution_major_version is version('7', '>=')) or 
        ansible_distribution == 'Fedora' or 
        (ansible_distribution == 'Ubuntu' and ansible_distribution_version is version('15.04', '>=')) or 
        (ansible_distribution == 'Debian' and ansible_distribution_version is version('8', '>=')) or 
        ansible_os_family == 'Suse'
    ```
- Regarding the module's failure conditions, you were right, two of them were unnecessary. Since I’m still new to writing Ansible modules, I tend to prefer failure over returning an empty result.
- I removed **`is_systemd`** and instead introduced a simple **`systemctl --version`** check within a **`try`**, failing if the command execution encounters an error.

    ```python
        systemctl_bin = module.get_bin_path('systemctl', required=True)
        try:
            run_command(module, [systemctl_bin, '--version'])
        except Exception as e:
            module.fail_json(msg="Failed to run systemctl: {0}".format(str(e)))
    ```
- I also removed the if condition on **`unitname`** (previously lines **270-271**).
- I renamed **`add_properties`** to **`extract_unit_properties`**. Additionally, **`fact`** is no longer passed to the function; instead, as you suggested, it is updated using **`fact.update`**.

    ```python
    fact.update(extract_unit_properties(unit_data, full_props))
    ```
    ```python
    def extract_unit_properties(unit_data, prop_list):
        extracted = {}
        for prop in prop_list:
            key = prop.lower()
            if key in unit_data:
                extracted[key] = unit_data[key]
        return extracted
    ```

russoz commented on plugins/modules/systemd_info.py:
```diff
@@ -0,0 +1,289 @@
+  - This module gathers info about systemd units (services, targets, sockets, mount).
+  - It runs C(systemctl list-units) (or processes selected units) and collects properties
+    for each unit using C(systemctl show).
+  - Even if a unit has a loadstate of "not-found" or "masked", it is returned,
```
Probably better as:
```suggestion
  - Even if a unit has a `loadstate` of `not-found` or `masked`, it is returned,
```

russoz replied:
Yup, my bad, I was doing that one in a rush. Sorry.

russoz commented on plugins/modules/systemd_info.py:
```diff
@@ -0,0 +1,289 @@
+  - It runs C(systemctl list-units) (or processes selected units) and collects properties
+    for each unit using C(systemctl show).
+  - Even if a unit has a loadstate of "not-found" or "masked", it is returned,
+    but only with the minimal properties ( name, loadstate, activestate, substate).
```
Editorial nit picking
```suggestion
    but only with the minimal properties (name, loadstate, activestate, substate).
```

russoz commented on plugins/modules/systemd_info.py:
```diff
@@ -0,0 +1,289 @@
+options:
+  unitname:
+    description:
+      - List of unit names to process.
```
I believe it will process all units (of those types) if this is not specified. If indeed so, do write that in the docs for the users.

NomakCooper replied:
For now, I have added an additional note in the description.
```yaml
- It supports .service, .target, .socket, and .mount units type.
```

russoz commented on plugins/modules/systemd_info.py:
```diff
@@ -0,0 +1,289 @@
+'''
+
+EXAMPLES = r'''
+# Gather info for all systemd services, targets, sockets and mount
```
Please add a YAML document marker:
```suggestion
---
# Gather info for all systemd services, targets, sockets and mount
```

russoz commented on plugins/modules/systemd_info.py:
```diff
@@ -0,0 +1,289 @@
+      - systemd-journald.service
+      - systemd-journald.socket
+      - sshd-keygen.target
+      - -.mount
```
Is this intentional? What is the meaning of `-. mount`?

NomakCooper replied:
The inclusion of the value **`-.mount`** in the example is intentional.

The systemd unit **`-.mount`** represents the root volume.

I included it in the example to ensure users know that **`.mount`** units beginning with the **`-`** character are supported.

russoz replied:
Thanks for clarifying

russoz commented on plugins/modules/systemd_info.py:
```diff
@@ -0,0 +1,289 @@
+
+
+def run_command(module, cmd):
+    rc, stdout, stderr = module.run_command(cmd, use_unsafe_shell=True)
+    if rc != 0:
+        module.fail_json(msg="Command '{0}' failed: {1}".format(" ".join(cmd), stderr.strip()))
```
Two issues here, that I missed before:
* why `use_unsafe_shell=True`?
* instead of doing that rc validation you can simply:

```suggestion
    rc, stdout, stderr = module.run_command(cmd, check_rc=True)
```

NomakCooper replied:
I used **`use_unsafe_shell=True`** as specified in the [Ansible documentation](https://docs.ansible.com/ansible/latest/dev_guide/developing_modules_best_practices.html).

[Module Security](https://docs.ansible.com/ansible/latest/dev_guide/developing_modules_best_practices.html#id11)
- **If you must use the shell, you must pass `use_unsafe_shell=True` to `module.run_command`.**

russoz replied:
> I used **`use_unsafe_shell=True`** as specified in the [Ansible documentation](https://docs.ansible.com/ansible/latest/dev_guide/developing_modules_best_practices.html).
> 
> [Module Security](https://docs.ansible.com/ansible/latest/dev_guide/developing_modules_best_practices.html#id11)
> - **If you must use the shell, you must pass `use_unsafe_shell=True` to `module.run_command`.**

The key thing being "if you must", which is not the case in your module. Nit sude how clear this is to you, the shell it mentions is /bin/bash (or a similar alternative), and it is only required if you need to use one of those shell features like `|`, or `>/somefile`, which is not the case. This module requires simply running a command, hence it should not set `use_unsafe_shell`.

NomakCooper replied:
I misunderstood the documentation. I can **`remove use_unsafe_shell=True`** and leave only **`check_rc=True`**.

russoz reviewed (commented):
Hi @NomakCooper thanks for the adjustments. I have a couple of additional comments, but I re iewed from the phone so I may have easily made some mistake. Take it with a grain of salt.

NomakCooper commented on plugins/modules/systemd_info.py:
```diff
@@ -0,0 +1,289 @@
+
+
+def extract_unit_properties(unit_data, prop_list):
+    extracted = {}
+    for prop in prop_list:
+        key = prop.lower()
+        if key in unit_data:
+            extracted[key] = unit_data[key]
+    return extracted
```
Your approach achieves the same result but in a more refined and elegant way. 😃

NomakCooper commented on plugins/modules/systemd_info.py:
```diff
@@ -0,0 +1,289 @@
+---
+module: systemd_info
+short_description: Gather C(systemd) unit info
+description:
```
Like this?

```yaml
extends_documentation_fragment:
  - community.general.attributes
  - community.general.attributes.info_module
```
Or should I add a new dedicated document for systemd with additional documentation?

like:

```yaml
extends_documentation_fragment:
  - community.general.systemd.documentation
  - community.general.attributes
  - community.general.attributes.info_module
```

The current documentation already includes all the necessary information about the options.
For the result values, I preferred to include an example, as it is much more illustrative.

russoz replied:
The very first option, exactly that.

NomakCooper replied:
In addition to adding **`community.general.attributes`** and **`community.general.attributes.info_module`**, I created a new **`community.general.systemd.documentation`**.

Should I leave it as is or remove it? I thought of adding it to provide as much information as possible.

NomakCooper commented:
I have made some changes based on the suggestions.

- Corrected the syntax in **`DOCUMENTATION`**.
https://github.com/ansible-collections/community.general/blob/4bf73d0f2705fd30b07b9970b3b9dc2595c63940/plugins/modules/systemd_info.py#L19

- Added a new line in the **`unitname`** description.
https://github.com/ansible-collections/community.general/blob/4bf73d0f2705fd30b07b9970b3b9dc2595c63940/plugins/modules/systemd_info.py#L24-L28
- Added **`extends_documentation_fragment`** and created the new **`plugins/doc_fragments/systemd.py`** with additional notes that handle the basic options and properties.
https://github.com/ansible-collections/community.general/blob/4bf73d0f2705fd30b07b9970b3b9dc2595c63940/plugins/doc_fragments/systemd.py#L14-L18
https://github.com/ansible-collections/community.general/blob/4bf73d0f2705fd30b07b9970b3b9dc2595c63940/plugins/doc_fragments/systemd.py#L29-L35
- Modified the **`run_command`** function to use **`use_unsafe_shell`** and **`check_rc`**.
https://github.com/ansible-collections/community.general/blob/4bf73d0f2705fd30b07b9970b3b9dc2595c63940/plugins/modules/systemd_info.py#L121-L123
According to the **[Ansible documentation](https://docs.ansible.com/ansible/latest/dev_guide/developing_modules_best_practices.html)**, using **`use_unsafe_shell`** is mandatory with **`module.run_command`**.
- Modified the **`extract_unit_properties`** function.
https://github.com/ansible-collections/community.general/blob/4bf73d0f2705fd30b07b9970b3b9dc2595c63940/plugins/modules/systemd_info.py#L159-L164

felixfontein commented:
> * Modified the **`run_command`** function to use **`use_unsafe_shell`** and **`check_rc`**.
>       https://github.com/ansible-collections/community.general/blob/4bf73d0f2705fd30b07b9970b3b9dc2595c63940/plugins/modules/systemd_info.py#L121-L123

I think you misunderstood that documenation. You only need it if you really need a shell. In 99% of all cases you don't need a shell. Using a shell is quite dangerous in general, so if there is no absolute need to use it, it should not be used.

Regarding the systemd docs fragment: I don't think that makes sense here.

felixfontein commented on plugins/modules/systemd_info.py:
```diff
@@ -0,0 +1,292 @@
+units:
+  description: Dictionary of systemd unit info keyed by unit name.
+  returned: success
+  type: dict
```
```suggestion
  type: dict
  contains:
    ... I think it makes sense to document the standard fields ...
```

NomakCooper replied:
I can try to document all the base properties of units, although I think there might be a problem with the .mount **`type`** property, as it might conflict with the syntax of the sanity test.

NomakCooper replied:
I have added the documentation for all the basic properties.

felixfontein commented on plugins/modules/systemd_info.py:
```diff
@@ -0,0 +1,292 @@
+      - sshd-keygen.target
+      - -.mount
+    extra_properties:
+        - Description
```
```suggestion
      - Description
```

felixfontein commented on plugins/modules/systemd_info.py:
```diff
@@ -0,0 +1,292 @@
+  - It runs C(systemctl list-units) (or processes selected units) and collects properties
+    for each unit using C(systemctl show).
+  - Even if a unit has a V(loadstate) of V(not-found) or V(masked), it is returned,
+    but only with the minimal properties (name, loadstate, activestate, substate).
```
```suggestion
    but only with the minimal properties (V(name), V(loadstate), V(activestate), V(substate)).
```

felixfontein commented on plugins/modules/systemd_info.py:
```diff
@@ -0,0 +1,292 @@
+  - Even if a unit has a V(loadstate) of V(not-found) or V(masked), it is returned,
+    but only with the minimal properties (name, loadstate, activestate, substate).
+  - When O(unitname) and O(extra_properties) are used, the module first checks if the unit exists,
+    then check if properties exist. If not, module fail.
```
```suggestion
    then check if properties exist. If not, the module fails.
```

felixfontein commented on plugins/modules/systemd_info.py:
```diff
@@ -0,0 +1,292 @@
+  unitname:
+    description:
+      - List of unit names to process.
+      - It supports .service, .target, .socket, and .mount units type.
```
```suggestion
      - It supports C(.service), C(.target), C(.socket), and C(.mount) units type.
```

felixfontein commented on plugins/modules/systemd_info.py:
```diff
@@ -0,0 +1,292 @@
+
+def unit_exists(module, systemctl_bin, unit):
+    if unit.startswith("-"):
+        cmd = [systemctl_bin, "show", "-p", "LoadState", "--", unit]
```
Why not always use this form? (Same in the other places.)

felixfontein replied:
Generally CLI programs allow `--` to denote "from now on there will only be positional arguments". Not every CLI program supports it, but many do. It's usually better to put `--` in there if you're just adidng more positional arguments.

NomakCooper replied:
Indeed, I have performed some tests and it works correctly as it should. 
In the last commit, I replaced the checks to use the **`--`** syntax by default.

felixfontein commented on plugins/modules/systemd_info.py:
```diff
@@ -0,0 +1,292 @@
+    default: []
+  extra_properties:
+    description:
+      - Additional properties to retrieve (appended to the default ones).
```
```suggestion
      - Additional properties to retrieve (appended to the default ones).
      - Note that all property names are converted to lower-case.
```

NomakCooper commented:
@felixfontein @russoz Thanks for the help and suggestions.

Regarding **`use_unsafe_shell`**, I had misunderstood the documentation. 

Latest changes:

- Removed **`use_unsafe_shell=True`**
- Removed documentation **`plugins/doc_fragments/systemd.py`**
- Added a new line to the **`description`** of **`unitname`**
https://github.com/ansible-collections/community.general/blob/168c8cb16cfd2d62416c788192db279f158ffb55/plugins/modules/systemd_info.py#L25-L29

NomakCooper commented:
- Corrected the documentation syntax.
- Changes the functions to use the **`systemctl show`** command with **`--`** characters by default.
- Updated the **`return`** documentation with the specifications of all the base properties.
https://github.com/ansible-collections/community.general/blob/b4c052186a8b1f305af9c86dcdc0e8d01d0bd115/plugins/modules/systemd_info.py#L68-L148

felixfontein commented:
I have some more docs suggestions:
```diff
diff --git a/plugins/modules/systemd_info.py b/plugins/modules/systemd_info.py
index f550b813d..a5e03a228 100644
--- a/plugins/modules/systemd_info.py
+++ b/plugins/modules/systemd_info.py
@@ -16,8 +16,8 @@ description:
   - This module gathers info about systemd units (services, targets, sockets, mount).
   - It runs C(systemctl list-units) (or processes selected units) and collects properties
     for each unit using C(systemctl show).
-  - Even if a unit has a V(loadstate) of V(not-found) or V(masked), it is returned,
-    but only with the minimal properties (V(name), V(loadstate), V(activestate), V(substate)).
+  - Even if a unit has a RV(units[].loadstate) of V(not-found) or V(masked), it is returned,
+    but only with the minimal properties (RV(units[].name), RV(units[].loadstate), RV(units[].activestate), RV(units[].substate)).
   - When O(unitname) and O(extra_properties) are used, the module first checks if the unit exists,
     then check if properties exist. If not, the module fails.
 version_added: "10.4.0"
@@ -66,7 +66,9 @@ EXAMPLES = r'''
 
 RETURN = r'''
 units:
-  description: Dictionary of systemd unit info keyed by unit name.
+  description:
+    - Dictionary of systemd unit info keyed by unit name.
+    - Additional fields will be returned depending on the value of O(extra_properties).
   returned: success
   type: dict
   elements: dict
@@ -79,24 +81,37 @@ units:
     loadstate:
       description:
         - The state of the unit's configuration load.
-        - Either V(loaded), V(not-found), V(masked).
       returned: always
       type: str
       sample: loaded
+      choices:
+        - loaded
+        - not-found
+        - masked
     activestate:
       description:
         - The current active state of the unit.
-        - Either V(active), V(inactive), V(failed).
       returned: always
       type: str
       sample: active
+      choices:
+        - active
+        - inactive
+        - failed
     substate:
       description:
         - The detailed sub state of the unit.
-        - Either V(running), V(dead), V(exited), V(failed), V(listening), V(active), V(mounted).
       returned: always
       type: str
       sample: running
+      choices:
+        - running
+        - dead
+        - exited
+        - failed
+        - listening
+        - active
+        - mounted
     fragmentpath:
       description: Path to the unit's fragment file.
       returned: always except for C(.mount) units.
@@ -105,17 +120,21 @@ units:
     unitfilepreset:
       description:
         - The preset configuration state for the unit file.
-        - Either V(enabled), V(disabled).
       returned: always except for C(.mount) units.
       type: str
       sample: disabled
+      choices:
+        - enabled
+        - disabled
     unitfilestate:
       description:
         - The actual configuration state for the unit file.
-        - Either V(enabled), V(disabled).
       returned: always except for C(.mount) units.
       type: str
       sample: enabled
+      choices:
+        - enabled
+        - disabled
     mainpid:
       description: PID of the main process of the unit.
       returned: only for C(.service) units.
```
Besides that, I think it's good.

NomakCooper commented:
Thanks @felixfontein 

- For the documentation changes, I wanted to use **`RV(...)`**, but I wasn’t sure about the correct syntax to use.

- Regarding the basic properties descriptions, I didn’t know that **`choices`** could be used. I thought it was only for documenting **`options`**. Additionally, the values I described are not the only possible values that exist in **`systemd`**.

For example, **`ActiveState`** could also be **`activating`**, **`deactivating`**, **`maintenance`**, **`reloading`**, or **`refreshing`**.

I included only the most common ones in the documentation since knowing all the possible values is quite difficult.

felixfontein commented:
> For example, **`ActiveState`** could also be **`activating`**, **`deactivating`**, **`maintenance`**, **`reloading`**, or **`refreshing`**.

But you wrote `Either V(active), V(inactive), V(failed).`. That means that it can be only one of the three values, and nothing else. If there are other possibilities, you should write that, like `The most common values are V(active), V(inactive), or V(failed), but other values are possible as well.`.

NomakCooper commented:
> But you wrote `Either V(active), V(inactive), V(failed).`. That means that it can be only one of the three values, and nothing else. If there are other possibilities, you should write that, like `The most common values are V(active), V(inactive), or V(failed), but other values are possible as well.`.

Sorry, my fault.
So I will replace the description lines "**Either...**" with "**most common values**" and not use **`choices`**.

NomakCooper commented:
I have updated the documentation. 

- Replaced the minimal properties in main description with the **`RV(...)`** values.
- **`RV(units[].prop)`** fail sanity test because units is a dict object, not a list.
    
> ```
> plugins/modules/systemd_info.py:0:0: DOCUMENTATION -> description[3]: RV(units[].activestate): return value name 'units[].activestate' refers to dictionary units with `[]`, which is only allowed for the last part
> plugins/modules/systemd_info.py:0:0: DOCUMENTATION -> description[3]: RV(units[].loadstate): return value name 'units[].loadstate' refers to dictionary units with `[]`, which is only allowed for the last part
> plugins/modules/systemd_info.py:0:0: DOCUMENTATION -> description[3]: RV(units[].loadstate): return value name 'units[].loadstate' refers to dictionary units with `[]`, which is only allowed for the last part
> plugins/modules/systemd_info.py:0:0: DOCUMENTATION -> description[3]: RV(units[].name): return value name 'units[].name' refers to dictionary units with `[]`, which is only allowed for the last part
> plugins/modules/systemd_info.py:0:0: DOCUMENTATION -> description[3]: RV(units[].substate): return value name 'units[].substate' refers to dictionary units with `[]`, which is only allowed for the last part
> ```

- Replaced **`RV(units[].prop)`** with **`RV(units.prop)`**
- For the basic properties, since not all possible values are listed, I did not use **`choices`**. Instead, I modified the description to make it clear that these are just **the most common results and other values are possible**.

russoz commented on plugins/modules/systemd_info.py:
```diff
@@ -0,0 +1,363 @@
+        supports_check_mode=True
+    )
+
+    systemctl_bin = module.get_bin_path('systemctl', required=True)
+    try:
+        run_command(module, [systemctl_bin, '--version'])
```
Wondering if maybe the `get_bin_path()` would not be better placed inside the `try` block - as it runs with `required=True` it may pop up an exception (on systems that do not have `systemctl` installed), and in that acse it might be more interesting to handle that exception with the `module.fail_json()` call, as in line 279.

NomakCooper replied:
Moving **`get_bin_path()`** inside the try block makes sense because it definitely adds value to the check before executing all the collection operations, but it’s not possible to insert a dedicated **`module.fail_json`**.

I don't think it’s a good idea to combine the try blocks.

To handle a dedicated **`module.fail_json`**, I could add a try block beforehand and then execute a second try block for the systemctl command.

sample:
```python
try:
    systemctl_bin = module.get_bin_path('systemctl', required=True)
except Exception as e:
    module.fail_json(msg="systemctl binary not found: {0}".format(str(e)))

try:
    run_command(module, [systemctl_bin, '--version'])
except Exception as e:
    module.fail_json(msg="Failed to run systemctl --version: {0}".format(str(e)))
```

felixfontein replied:
@russoz the free function `get_bin_path` does that, `AnsibleModule.get_bin_path` calls `AnsibleModule.fail_json` instead.

russoz commented on plugins/modules/systemd_info.py:
```diff
@@ -0,0 +1,363 @@
+                continue
+
+            props = base_properties.get(category, [])
+            full_props = list(set(props + state_props))
```
Not that it makes a huge difference, but I believe the `list()` wrapper is not needed.

NomakCooper replied:
Using **`list()`** is not strictly necessary for execution; it's just a way to ensure the result.

russoz reviewed (commented):
Couple of minor comments, other than that it LGTM

NomakCooper commented:
> Couple of minor comments, other than that it LGTM

Thanks @russoz 

- Added a new **`try`** block for **`get_bin_path()`** with a dedicated **`module.fail_json.`**

    https://github.com/ansible-collections/community.general/blob/3a4019914a8c48b2dbffd4ad29532b85eb2eeb12/plugins/modules/systemd_info.py#L275-L283
- Removed **`list()`** from the properties **`set()`**.

    https://github.com/ansible-collections/community.general/blob/3a4019914a8c48b2dbffd4ad29532b85eb2eeb12/plugins/modules/systemd_info.py#L332
    
    https://github.com/ansible-collections/community.general/blob/3a4019914a8c48b2dbffd4ad29532b85eb2eeb12/plugins/modules/systemd_info.py#L350

felixfontein commented on plugins/modules/systemd_info.py:
```diff
@@ -0,0 +1,367 @@
+    loadstate:
+      description:
+        - The state of the unit's configuration load.
+        - The most common values are V(loaded), V(not-found), V(masked), but other values are possible as well.
```
Same for the others.
```suggestion
        - The most common values are V(loaded), V(not-found), and V(masked), but other values are possible as well.
```

NomakCooper replied:
Descriptions have been fixed

felixfontein commented on plugins/modules/systemd_info.py:
```diff
@@ -0,0 +1,367 @@
+        supports_check_mode=True
+    )
+
+    try:
+        systemctl_bin = module.get_bin_path('systemctl', required=True)
+    except Exception as e:
+        module.fail_json(msg="systemctl binary not found: {0}".format(str(e)))
```
`AnsibleModule.get_bin_path` with `required=True` calls `module.fail_json` directly. There's no exception to catch.
```suggestion
    systemctl_bin = module.get_bin_path('systemctl', required=True)
```

NomakCooper replied:
Removed try block

felixfontein commented on plugins/modules/systemd_info.py:
```diff
@@ -0,0 +1,367 @@
+    except Exception as e:
+        module.fail_json(msg="systemctl binary not found: {0}".format(str(e)))
+
+    try:
+        run_command(module, [systemctl_bin, '--version'])
+    except Exception as e:
+        module.fail_json(msg="Failed to run systemctl: {0}".format(str(e)))
```
`AnsibleModule.run_command` by default also does not raise exceptions, but calls `fail_json`.
```suggestion
    run_command(module, [systemctl_bin, '--version'])
```

NomakCooper replied:
Removed try block

NomakCooper commented:
@felixfontein 

- Updated the descriptions as you suggested
- Removed the try blocks for **`get_bin_path`** and **`run_command`**

russoz reviewed (approved):
LGTM

felixfontein commented on tests/integration/targets/systemd_info/tasks/main.yml:
```diff
@@ -0,0 +1,25 @@
+    ansible_distribution == 'Fedora' or 
+    (ansible_distribution == 'Ubuntu' and ansible_distribution_version is version('15.04', '>=')) or 
+    (ansible_distribution == 'Debian' and ansible_distribution_version is version('8', '>=')) or 
+    ansible_os_family == 'Suse'
```
Is there a reason that the tests aren't run for Arch Linux?

NomakCooper replied:
No, honestly I added the checks on **`ansible_distribution`** because I don't really know all the types of hosts on which the integration test is run.

To ensure that the test is executed on all hosts regardless of version, I could replace the **`ansible_distribution`** checks with a simple check on **`ansible_service_mgr`**:

```yaml
when: ansible_service_mgr == 'systemd'
```

or I could simply add Archlinux :
```yaml
ansible_distribution == 'Archlinux'
```

felixfontein replied:
I'd probably simply add `ansible_distribution == 'Archlinux'`.

NomakCooper replied:
Added **`ansible_distribution == 'Archlinux'`**

felixfontein commented on tests/integration/targets/systemd_info/tasks/tests.yml:
```diff
@@ -0,0 +1,107 @@
+# SPDX-License-Identifier: GPL-3.0-or-later
+
+- name: Gather all units from shell
+  ansible.builtin.shell: systemctl list-units --no-pager --type service,target,socket,mount --all --plain --no-legend
```
```suggestion
  ansible.builtin.command: systemctl list-units --no-pager --type service,target,socket,mount --all --plain --no-legend
```

NomakCooper replied:
Replaced **`shell`** with **`command`**

felixfontein commented on tests/integration/targets/systemd_info/tasks/tests.yml:
```diff
@@ -0,0 +1,107 @@
+    label: "{{ item[0] }}"
+
+- name: Gather systemd-journald.service properties from shell
+  ansible.builtin.shell: systemctl show systemd-journald.service -p Id,LoadState,ActiveState,SubState,FragmentPath,MainPID,ExecMainPID,UnitFileState,UnitFilePreset,Description,Restart
```
```suggestion
  ansible.builtin.command: systemctl show systemd-journald.service -p Id,LoadState,ActiveState,SubState,FragmentPath,MainPID,ExecMainPID,UnitFileState,UnitFilePreset,Description,Restart
```

NomakCooper replied:
Replaced **`shell`** with **`command`**

felixfontein commented:
@NomakCooper thanks for your contribution!
@russoz thanks for reviewing!
